### ISSUE TITLE
Look for some guidance on batched block sparse operations

### ISSUE THREAD

**Issue by jerrybai1995** [getkeops/keops#80]
Hello,

I'm looking for some guidance on the block-sparse operation that we can apply in batch. In particular, what I want to do can be summarized in the following toy setting:
1) I have a "coarse-level" matrix `K`, say of dimension TxT. This is computed in the original pytorch.
2) I also have two fine-level tensors `x`, `y` of dimension `Nxm` both, for N > T. There is an alignment between T and N (i.e., I can group the N elements into T clusters, which form the "blocks"). 
3) I would like to compute `x @ y.t()`, which is `NxN`, but only in specific regions where `K > K.mean()`. 

As an example, here is how I would do it in pytorch (for simplicity, let's say T=3, N=7, m=4):
```python
batch_size = 2
T = 3
N = 7
m = 4
row_alignment = torch.tensor([[0,0,1,1,1,1,2],[0,1,1,2,2,2,2]]).view(batch_size, N, 1).long().repeat(1, 1, N)    # (bsz x N x N)
col_alignment = torch.tensor([[0,0,0,0,1,1,2],[0,1,1,1,1,1,2]]).view(batch_size, 1, N).long().repeat(1, N, 1)    # (bsz x N x N)
K = torch.randn(batch_size, T, T)
x = torch.randn(batch_size, N, m)
y = torch.randn(batch_size, N, m)

# First, filter out the "hot areas" that we want to compute at the finer-level by a threshold (I'm using the mean here)
K_filtered = K > (K.mean(2).mean(1))[:,None,None]

# We then stretch K to NxN according to the alignment
K_ = K_filtered[torch.arange(batch_size)[:,None,None], row_alignment, col_alignment]
result = x @ y.transpose(1,2) * (K_.float())
```
Essentially, only areas of the `NxN` result that correspond to regions of `K` that is larger than the threshold are eventually kept. I think this scenario fits the block sparse computation of keops very well (with the tutorial at https://www.kernel-operations.io/keops/_auto_examples/pytorch/plot_grid_cluster_pytorch.html).

However, it seems to me that this is rather challenging to do with a batch dimension. Specifically, what would be a suggested/the best way to apply the `range_ij` function to tensors of shape `(bsz x M x N)`? Is there any advice on how I should conduct the aforementioned operation in keops?

Thanks a lot in advance!

**Comment by jeanfeydy** [comment]
Hi @jerrybai1995 ,

Thanks for your detailed question!
Using KeOps, creating a symbolic LazyTensor that corresponds to your desired `result` should not be too difficult. I would do this in four steps:

1. **Reshape** the data arrays `x` and `y` as PyTorch tensors of shape `(B*N, m)`, where `B` is the batchsize.

2. Create a symbolic tensor `result` that encodes the **full** Gram matrix `x @ y.t()` of shape `(B*N, B*N)`. This can be done using the dot-product formula `(x_i | y_j)`.

3. For each of the `B` matrices `K[b]` of shape `(T, T)`, compute a **block-sparse truncation mask** of size `(N, N)` using the [from_matrix(...)](http://www.kernel-operations.io/keops/python/api/pytorch/Cluster.html#pykeops.torch.cluster.from_matrix) helper. The masks will be encoded in `B` 6-uples of integer Tensors, the "ranges".

4. **Fuse** the `B` truncation masks above into a single truncation mask of shape `(B*N, B*N)`, and use it as the `result.ranges` attribute. This can be done with simple PyTorch operations on the "ranges" computed at step 3, using **concatenations** and **offsets by integers**.

The last point is fairly low-level, and relies on the precise specifications of the `ranges` arrays described in the [documentation](http://www.kernel-operations.io/keops/python/api/pytorch/Genred_torch.html#pykeops.torch.Genred.__call__): I can help you with this if you believe that KeOps is well suited to your problem. A minimal working example could be turned into a good tutorial for advanced users!

Note, however, that `result` would be a symbolic LazyTensor and not a dense PyTorch matrix. If you intend to perform **reductions** on `result` (e.g. a sum or a logsumexp for an attention layer), that would be perfect - but you may have something else in mind. 

Would this be fine for you?
And in practice, which values for `N`, `m`, `T` or `B` would you like to handle?
(Depending on the orders of magnitude involved, we may favor one method over the others.)

Best regards,
Jean

**Comment by jerrybai1995** [comment]
Hi @jeanfeydy,

Thank you so much for the prompt response and the detailed explanation! A few follow-up questions on your response:

0) (Not really a question) My intended values are `N=500`, `m=64`, `T=70` and `B~=25`. However, depending on the threshold I set, the eventual `NxN` matrix could be really sparse, which is what motivated me to look into keops.

1) I don't quite understand the rationale behind creating the full Gram matrix. Is it only a way to reshape the tensor to 2D (in which we actually only need the diagonal blocks)?

2) Regarding the usage of `from_matrix` that you mentioned in step 3, should I use the filtered `K[b]` as `keep`, and produce the corresponding `range_i` and `range_j` using the `row_alignment` and `col_alignment` (see my code above) by converting these original alignment indices to start- and end-indices of each cluster (which have shape `Tx2`)? 

3) How exactly should I fuse the truncation masks? Because this is intended to be a batched operation, the way I see it is that I only need the diagonal blocks of the full Gram matrix, which is where these `B` "ranges" need to be set. But I'm not entirely certain what I need to do to concatenate them. Do you have a pointer to how to this?

Re - your question on what happens to `result`: eventually this `NxN` (sparse) matrix will go through a non-linearity and then be dotted with a `Nxr` tensor. My worry though, is that in order for `result` to go through a non-linear function, it can't remain in the symbolic LazyTensor form, correct?

Again, thanks so much for the detailed guidance!

**Comment by jeanfeydy** [comment]
Hi @jerrybai1995 ,

You're welcome! To answer your points:

0. As explained in [our documentation](http://www.kernel-operations.io/keops/python/sparsity.html), block-sparsity leverages the [group-wise structure](http://www.kernel-operations.io/keops/autodiff_gpus/what_is_a_gpu.html) of CUDA computations. In practice, speed-ups against a brute-force KeOps implementation are only achieved if all threads in a CUDA block are kept busy by the [tiles](http://www.kernel-operations.io/keops/_auto_examples/pytorch/plot_grid_cluster_pytorch.html) specified through the `ranges` attribute of a LazyTensor. If the ratio `N / T` is much smaller than 100, you may consider using a **coarser block-sparsity mask** (i.e. decrease T): somewhat disappointingly, on the GPU, schemes that are "too sparse" are often outperformed by semi-bruteforce approaches.


1. Indeed: batch support in KeOps is little more than syntactic sugar around a **block-diagonal reduction**. The relevant `ranges` are created on-the-fly by the [C++ code](https://github.com/getkeops/keops/blob/082dce50d77363239164a588c70c0e2af05c7a6c/keops/binders/switch.h#L74). As you plan to do something a bit more subtle, the simplest way of doing so with the current API is to create your own reduction ranges "by hand".


2. Exactly. For reference, here is a code snippet that turns "batch vectors" (i.e. vectors of integer labels of the form `[0, 0, ..., 0, 1, ..., 1, ..., n-1, ..., n-1]`) into a block-diagonal sparsity mask for KeOps. I wrote it for integration with [PyTorch_geometric's batch mode](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html#mini-batches) and will soon add a helper routine in the main repository.

```python
def ranges_slices(batch):
    """Helper function for the diagonal ranges function."""
    Ns = batch.bincount()
    indices = Ns.cumsum(0)
    ranges = torch.cat((0 * indices[:1], indices))
    ranges = torch.stack((ranges[:-1], ranges[1:])).t().int().contiguous().to(batch.device)
    slices = (1 + torch.arange(len(Ns))).int().to(batch.device)
    
    return ranges, slices


def diagonal_ranges(batch_i = None, batch_j = None):
    """Encodes the block-diagonal structure associated to a batch vector."""
    
    if batch_x is None and batch_j is None: return None
    
    ranges_i, slices_i = ranges_slices(batch_i)
    ranges_j, slices_j = ranges_slices(batch_j)
    
    return ranges_i, slices_i, ranges_j, ranges_j, slices_j, ranges_i 
```


3. Indeed: with the current API, the simplest way of working is to "unwrap" the batch dimensions prior to a KeOps reduction. You would then perform a reduction on a LazyTensor of shape `(B*N, B*N)` and reshape the result as a PyTorch tensor of shape `(B, N, ...)` afterwards. If the example above is not sufficient for you, I could do this a bit later. Going forward, the best API would be to let `from_matrix` support batch processing and perform this "fusion" automatically!


I agree with you that the block-sparse mode is not very well documented yet: I use it extensively for my work on optimal transport, but did not expect many people to use it beyond straightforward batch processing. You are probably one of the very first users of KeOps who went this far into the documentation :-)


4. I see: there should be no problem then. In practice, your code will probably look like:

```python
import torch
from pykeops.torch import LazyTensor

# Toy data:
B, N, m, r = 25, 500, 64, 32
x = torch.rand(B, N, m)
y = torch.rand(B, N, m)
b = torch.rand(B, N, r)

# Encoding as LazyTensor:
x_i = LazyTensor( x.view(B*N, 1, m) )
y_j = LazyTensor( y.view(1, B*N, m) )

# Gram matrix:
result_ij = (x_i | y_j)  # (B*N, B*N)

# Block-sparse truncation:
result_ij.ranges = ...  # Use a fused mask, or the block-diagonal ranges above.

# Non-linearity:
R_ij = result_ij.relu()  # (B*N, B*N)

# Matrix-vector product:
output = (R_ij @ b.view(B*N, r)).view(B, N, r)

```


If you need a non-linearity that is not yet support by KeOps, feel free to create an issue: as discussed in the [documentation](http://www.kernel-operations.io/keops/formulas/math_operations.html) and illustrated [here](https://github.com/getkeops/keops/blob/082dce50d77363239164a588c70c0e2af05c7a6c/keops/core/formulas/maths/ReLu.h) for ReLU, implementing a new operator is a matter of minutes.


Note that before implementing the fusion of the ranges, I would first advise you to try a "semi-bruteforce" KeOps implementation with pure batch processing and no truncation: performance should not be too bad, as you intend to use a "small" value of `N=500`. If the library fits well within your codebase, we could then look into further optimization with a block-wise sparsity mask.
What do you think?

Best regards,
Jean

**Comment by jeanfeydy** [comment]
I close the issue as there has been no activity for 2+ years now, but feel free to re-open if needed.
